cmd/jaeger/internal/integration/e2e/integration.go

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) 2024 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package e2e
+
+import (
+	"context"
+	"errors"
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	jaeger2otlp "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/jaeger"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/testbed/testbed"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/status"
+
+	"github.com/jaegertracing/jaeger/cmd/jaeger/internal"
+	"github.com/jaegertracing/jaeger/model"
+	"github.com/jaegertracing/jaeger/plugin/storage/integration"
+	"github.com/jaegertracing/jaeger/ports"
+	"github.com/jaegertracing/jaeger/proto-gen/api_v2"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+)
+
+var (
+	_ spanstore.Writer = (*spanWriter)(nil)
+	_ spanstore.Reader = (*spanReader)(nil)
+)
+
+// E2EStorageIntegration holds components for e2e mode of Jaeger-v2
+// storage integration test. The intended usage is as follows:
+//   - A specific storage implementation declares its own test functions
+//     (e.g. starts remote-storage).
+//   - In those functions, instantiates with e2eInitialize()
+//     and clean up with e2eCleanUp().
+//   - Then calls RunTestSpanstore.
+type E2EStorageIntegration struct {
+	integration.StorageIntegration
+	ConfigFile string
+
+	runner        testbed.OtelcolRunner
+	configCleanup func()
+}
+
+// e2eInitialize starts the Jaeger-v2 collector with the provided config file,
+// it also initialize the SpanWriter and SpanReader below.
+func (s *E2EStorageIntegration) e2eInitialize() error {
+	factories, err := internal.Components()
+	if err != nil {
+		return err
+	}
+
+	config, err := os.ReadFile(s.ConfigFile)
+	if err != nil {
+		return err
+	}
+	config = []byte(strings.Replace(string(config), "./cmd/jaeger/", "../../../", 1))
+
+	s.runner = testbed.NewInProcessCollector(factories)
```

Comment:
Personally I have a preference towards starting the actual jaeger-v2 binary because that is a much more real e2e test. From performance considerations I think it doesn't matter much, because starting the storage backend is the more expensive part anyway. We could even go further and run the binary as a Docker image, for an even deeper e2e testing which is still needed (otherwise how do we know the Docker image we built is actually working?).

It's ok to leave this for now, but I would suggest in the follow-ups to refactor this piece into a swappable component, so that we could have different modes of e2e testing.

Reply:
> Personally I have a preference towards starting the actual jaeger-v2 binary... We could even go further and run the binary as a Docker image,

Working on it!